internal/controller/postgrescluster/snapshots.go

```diff
@@ -133,49 +148,286 @@ func (r *Reconciler) reconcileVolumeSnapshots(ctx context.Context,
+	// For now, I am choosing the first instanceSet/instance in their arrays.
+	instanceSet := &cluster.Spec.InstanceSets[0]
+	instance := instances.bySet[instanceSet.Name][0].Runner
+
+	// How are we going to handle tablespace volumes? Feels like if they exist we need to provide them to
+	// the restore command otherwise they will break in any clones made from a snapshot.
+	pgtablespaces, err := r.reconcileTablespaceVolumes(ctx, cluster, instanceSet, instance, clusterVolumes)
+	if err != nil {
+		return errors.WithStack(err)
 	}
-	// Return error if primary pvc not found
-	if primaryPvc.Name == "" {
-		return nil, errors.New("Could not find primary's pgdata pvc. Cannot create volume snapshot.")
+
+	cmd := pgbackrest.DedicatedSnapshotVolumeRestoreCommand(pgdata, pgtablespaces, strings.Join(opts, " "))
+
+	// Create the volume resources required for the Postgres data directory.
+	dataVolumeMount := postgres.DataVolumeMount()
+	dataVolume := corev1.Volume{
+		Name: dataVolumeMount.Name,
+		VolumeSource: corev1.VolumeSource{
+			PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+				ClaimName: dedicatedSnapshotVolume.GetName(),
+			},
+		},
 	}
+	volumes := []corev1.Volume{dataVolume}
+	volumeMounts := []corev1.VolumeMount{dataVolumeMount}
+
+	_, configHash, err := pgbackrest.CalculateConfigHashes(cluster)
+	if err != nil {
+		return err
+	}
+
+	// A DataSource is required to avoid a nil pointer exception.
+	fakeDataSource := &v1beta1.PostgresClusterDataSource{RepoName: ""}
+
+	restoreJob := &batchv1.Job{}
+	instanceName := cluster.Status.StartupInstance
 
-	// generate VolumeSnapshot
-	snapshot, err := r.generateVolumeSnapshot(postgrescluster, primaryPvc,
+	if err := r.generateRestoreJobIntent(cluster, configHash, instanceName, cmd,
+		volumeMounts, volumes, fakeDataSource, restoreJob); err != nil {
+		return errors.WithStack(err)
+	}
+
+	// Attempt the restore exactly once. If the restore job fails, we prompt the user to investigate.
+	restoreJob.Spec.BackoffLimit = initialize.Int32(0)
+	restoreJob.Spec.Template.Spec.RestartPolicy = corev1.RestartPolicyNever
+
+	// Add pgBackRest configs to template.
+	pgbackrest.AddConfigToRestorePod(cluster, cluster, &restoreJob.Spec.Template.Spec)
+
+	// Add nss_wrapper init container and add nss_wrapper env vars to the pgbackrest restore container.
+	addNSSWrapper(
+		config.PGBackRestContainerImage(cluster),
+		cluster.Spec.ImagePullPolicy,
+		&restoreJob.Spec.Template)
+
+	addTMPEmptyDir(&restoreJob.Spec.Template)
+
+	restoreJob.Annotations[naming.PGBackRestBackupJobCompletion] = backupJob.Status.CompletionTime.String()
+	return errors.WithStack(r.apply(ctx, restoreJob))
+}
+
+// generateSnapshotOfDedicatedSnapshotVolume will generate a VolumeSnapshot of
+// the dedicated snapshot PersistentVolumeClaim and annotate it with the
+// provided backup job's UID.
+func (r *Reconciler) generateSnapshotOfDedicatedSnapshotVolume(
```

Comment:
:pick: Feels like the code in `generateVolumeSnapshot` should be in here.

Reply:
I kept it in its separate function so that it can be reused if we ever snapshot other volumes.